### ISSUE TITLE
OCULUS QUEST  working with sTEAM  VR¿?

### ISSUE THREAD

**Issue by chapeka** [polygraphene/ALVR#287]
May be possible a new ALVR releasse fro the Oculus quest¿??¿

**Comment by polygraphene** [comment]
I'm sorry that I have not respond to GitHub issues for a long time.

I have a plan to support Quest. But, don't expect too much.

**Comment by AltoRetrato** [comment]
I don't know how ALVR currently does video compression, but I wonder if something similar to [fixed foveated rendering](https://developer.oculus.com/documentation/mobilesdk/latest/concepts/mobile-ffr/) (as seen on Oculus Go) could be implemented - i.e., reducing resolution and/or quality at the edges of the frame. It would add complexity to the code, but could (in theory) reduce latency and/or improve image quality at the center.

**Comment by polygraphene** [comment]
@AltoRetrato ALVR uses H.264/H.265 compression provided by GPU manufacturer. I don't think FFR  can reduce latency on video compression, because the processing time does not rely on rendered image.
And you need RTX GPU to use FFR on PC.

**Comment by AltoRetrato** [comment]
@polygraphene  I don't mean to render the game image on the PC using FFR.

The idea is to reduce image size by 30% (or more) after it is rendered by discarding pixels in the periphery before it is compressed and sent to the headset.

**Comment by polygraphene** [comment]
Do you mean split image to multiple part according to density of FFR?
I can't predict the performance when splitting image and feeding it into encoder...
Splitting image also takes a time and encoder have limitation on how many images can be processed at the same time.

**Comment by AltoRetrato** [comment]
Splitting the original images into smaller ones and processing each one individually could be one way to do it, but it would be expensive.

Another way would be to "remap" the original image instead of splitting into multiple images. Think about making something like a texture atlas: it is a single image, with different regions being mapped to different portions of the screen at different densities.

**Comment by polygraphene** [comment]
FFR+"remap" is an interesting idea. That has low processing cost and moderate programming cost.
I want to try it when I have a time. But I'm busy at Daydream(and Quest) support on ALVR now. So I can't estimate when I can start it.

Thank you for providing good idea.

**Comment by polygraphene** [comment]
On GearVR and Go, I have published client by key distribution. So the users didn't need to enable developer mode to install client.

But unfortunately, Oculus said key distribution won't be available on Quest.
So all users have to install client by enabling developer mode.

**Comment by jc-langevin** [comment]
any ETA on your work for the oculus quest?

**Comment by peterthrees** [comment]
I really hope you get it done for quest, it would be amazing.
If i can help you with something, let me know. I'm a software developer for years, don't have any experience in hardware but I'm a fast learner :)

**Comment by MarcusSjolin** [comment]
I more than happily get involved as well, will have a quest by next week I presume, so will check the code out till then.

**Comment by polygraphene** [comment]
@jc-langevin 
My order has been postponed by Amazon JP. I can't believe it.
There is no ETA for my Quest.

@peterthrees 
I will publish pre-release version once it is done. I would appreciate if you can test and provide log information for debug.

**Comment by peterthrees** [comment]
@polygraphene:  Np, I can do that for sure , I'll keep eye on it when the pre-release is out.

**Comment by polygraphene** [comment]
@MarcusSjolin 
Thank you! Feel free to ask any question about my codes.

I heard ALVR has following issues on Quest now. (#288)
1. side-by-side was incorrect. (Resolution issue?)
2. Only one controller is usable.

I'm working on first issue now.

**Comment by FedericoRuiz1** [comment]
Hey! Thanks for doing this. Will it support 6dof?

**Comment by jc-langevin** [comment]
@FedericoRuiz1 look at #288

**Comment by XDRitz** [comment]
@polygraphene how can I help test your builda?

**Comment by DannyDan** [comment]
@polygraphene I also want to support here with beta testing.

**Comment by polygraphene** [comment]
@XDRitz @DannyDan Please try it. https://github.com/polygraphene/ALVR/releases/tag/v2.4.0-alpha1
I don't know whether controller works.
I want to know the resolution shown in ALVR.exe when client is connected.

**Comment by SpookyBear0** [comment]
@polygraphene I tested it out, everything works perfect except for that there are two screens (for vr) do you have any way to fix this?

PS: Both controllers work

**Comment by polygraphene** [comment]
Thank you for trying.
What do you mean by "two screen"? I want to see a screenshot.
And can you tell me the refresh rate and resolution shown below the device
name?

> Both controllers work
That's a good news. Are the all buttons working?


2019年5月24日(金) 4:08 SpookyBear0 <notifications@github.com>:

**Comment by SpookyBear0** [comment]
@polygraphene  I will check (I cant check the controllers bc of the "two screen" but I will send screenshot and stuff)

**Comment by SpookyBear0** [comment]
@polygraphene 72Hz and 2432x1344, I can't get  a picture bc its not working rn and other reasons but its like when you use a phone vr and the picture doubles. (example shown)
![br](https://user-images.githubusercontent.com/50897638/58280735-28ff7300-7d56-11e9-93d7-15d866a7d315.jpg)

**Comment by AnalogMan151** [comment]
I have tested it out as well. I also get double images. Like, it's doing stereoscopic correctly but the distance between the images is too far so it's double-vision. 
As for the controllers, only the right controller works for me and it shows in-game as a Vive controller. Only the trigger and grip buttons have function.

**Comment by SpookyBear0** [comment]
@polygraphene Yeah thats what I have too (video here to get an idea) https://drive.google.com/file/d/1E7T8p03kxQbyaaYIaekwpvKTlqx_X6Yu/view

**Comment by AnalogMan151** [comment]
Resolution reported by ALVR is 2432x1344 @ 72 Hz

Official res is 2560x1440@72. 
Also just found the controller settings, that should fix the buttons. Will report back.

**Comment by AnalogMan151** [comment]
> 
> 
> @polygraphene Yeah thats what I have too (video here to get an idea) https://drive.google.com/file/d/1E7T8p03kxQbyaaYIaekwpvKTlqx_X6Yu/view

File is not public.

**Comment by polygraphene** [comment]
Thank you for testing.
It is difficult to debug without device. My device will be delivered today.
So I will start to debug after arrival.
Please be patient.

2019年5月24日(金) 5:02 AnalogMan <notifications@github.com>:

**Comment by SpookyBear0** [comment]
Thanks for making this program @polygraphene

**Comment by AnalogMan151** [comment]
Question, are the controllers meant to show up as Vive or should they be identifying as Oculus Touch controllers?

**Comment by DannyDan** [comment]
Too bad I need an USB cable to connect to my Computer to test this too. :(
But I found another project:
https://support.riftcat.com/hc/en-us/articles/360005631080

**Comment by AnalogMan151** [comment]
@DannyDan What do you mean USB cable?

**Comment by AnalogMan151** [comment]
@polygraphene I almost want to say the images are reversed. When I close one eye what I expect to see in my left eye is what my right eye should be seeing and vice-versa. Is that even possible?
Also, @SpookyBear0 how did you get both controllers to work? Only one at a time will work for me. Either only the right or only the left or only the left (but thinks it's the right).

**Comment by SpookyBear0** [comment]
@AnalogMan151 Idk

**Comment by SpookyBear0** [comment]
@polygraphene Did you get your quest? If you did I hope you like it.

**Comment by Zetaphor** [comment]
@polygraphene @AnalogMan151 I'm also getting the reversed image problem. I was seeing the "Complete setup on PC" dialog telling me to do the SteamVR device setup. When I closed my right eye I saw the left side of the dialog, right eye saw the left. Seems like the video output was mapping to the wrong side

**Comment by Zetaphor** [comment]
> 
> 
> Question, are the controllers meant to show up as Vive or should they be identifying as Oculus Touch controllers?

@AnalogMan151 Yes, since SteamVR thinks it's seeing Vive wands. We're not going to be able to get a native touch model through Steam

**Comment by polygraphene** [comment]
I got Quest just now! 6DoF has greatly improved immersion compared with GearVR.

I confirmed doubling issue and controller issue.
Maybe it was caused by FOV? I'm trying to investigate the issues.

**Comment by DannyDan** [comment]
> @DannyDan What do you mean USB cable?

@AnalogMan151 
I need an Type C cable to copy the apk to the device.

**Comment by AnalogMan151** [comment]
@DannyDan Ah, true. But the same goes for the other project you linked.

**Comment by polygraphene** [comment]
> > Question, are the controllers meant to show up as Vive or should they be identifying as Oculus Touch controllers?
> 
> @AnalogMan151 Yes, since SteamVR thinks it's seeing Vive wands. We're not going to be able to get a native touch model through Steam

Do you mean we see Vive wands even when we use Oculus Rift(+Touch) in SteamVR?

**Comment by SweViver** [comment]
Indeed it looks like the image in each eye is reversed in the Quest, left image in right eye and vice versa. Is this possible to fix? Also any way to run this through USB? I cant find any option for that. It only works through Wifi here...

**Comment by polygraphene** [comment]
Doubling issue on eye images seems fixed now.
Currently, button assignment does not work for touch controller.
https://github.com/polygraphene/ALVR/releases/tag/v2.4.0-alpha2

**Comment by Woohaaa** [comment]
some wrong with "trackpad click" about a2 ver in quest!

**Comment by peterthrees** [comment]
I've added a post on reddit: https://www.reddit.com/r/OculusQuest/comments/bsfuba/vridge_experience_and_alvr_on_quest/

It seems that ALVR does it much better at the moment than riftcat. Great job so far! @polygraphene 

I'll try it out tonight and send you my feedback.

**Comment by Irastris** [comment]
Just gave Beat Saber a run on Expert+ using Alpha 2.

Unfortunately in really heavy areas with lots of controller movement, the stream drops a ton of frames, even on 5GHz WiFi. Expert and lower difficulty run near perfect, maybe 20ms of latency.

Very compelling work though, considering Quest is only 3 days old. I'd be extremely interested if ALVR implemented USB tethering as iVRy is making claims to have support for.

**Comment by FricoRico** [comment]
> Just gave Beat Saber a run on Expert+ using Alpha 2.
> 
> Unfortunately in really heavy areas with lots of controller movement, the stream drops a ton of frames, even on 5GHz WiFi. Expert and lower difficulty run near perfect, maybe 20ms of latency.
> 
> Very compelling work though, considering Quest is only 3 days old. I'd be extremely interested if ALVR implemented USB tethering as iVRy is making claims to have support for.

Can you confirm your Quest is actually getting a high quality connection to your network? Mine sometimes only reaches 173Mbps while I've seen it reach ~900Mbps usually. You can check by pressing on your connected WiFi network in the settings.

**Comment by Irastris** [comment]
> Can you confirm your Quest is actually getting a high quality connection to your network?

According to the Quest, I'm reaching a constant 866Mbps.

**Comment by moallan03** [comment]
i thought u couldnt use controllers with alpha 2? Or is that only the thumbstick?

**Comment by Irastris** [comment]
2 controller tracking and the triggers/grips work fine. It seems to be only the face buttons and joystick that are unresponsive. Since Beat Saber only requires the triggers for menu navigation, that was enough to get in-game for the test.

**Comment by moallan03** [comment]
Interesting, does anyone know how to uninstall the first alpha? I wanna download the second but i dont know how to delete it if a third alpha comes out

**Comment by Irastris** [comment]
`adb uninstall com.polygraphene.alvr`

**Comment by machenmusik** [comment]
FYI I was able to play In Death but custom input mapping was required... for that game to see a remapped grip press as the right trackpad press (teleport arrow) I had to make it send press, touch AND nonzero analog position.  Clearly alpha 2 was not satisfying all those conditions with the right stick...

**Comment by lvoogdt** [comment]
Wow, it works brilliantly here, looks very good with resolution at 150%. Around 60ms latency is reported, i am suprised at how playable it feels. 

What is needed to make the controller buttons working? Would it also be possible to simulate oculus controllers somehow in steamvr?

Great work, keep it up!

**Comment by SpookyBear0** [comment]
@polygraphene someone streaming ALVR: [https://www.twitch.tv/ragesaq](url)

**Comment by Irastris** [comment]
We *could* make SteamVR display the old CV1 controllers by modifying `ServerConfig.cs`. The main problem though is that the server currently maps the same rendermodel to both controllers, so you'd either have two left hands or two right hands. Making the server apply separate rendermodels is beyond my level of knowledge, unfortunately, so I will leave that to @polygraphene .

I'm currently digging through my Quest's filesystem to try and find the models for the new controllers. With some work they could be adapted to SteamVR's rendermodel system.

**Comment by PixlM** [comment]
are there any plans on supporting usb c  tethering as well?

**Comment by FingrMastr** [comment]
I use usb with a simple Ethernet adapter on Oculus Go, and this way reducing latency and being able to push bitrate (less artifacts). Real usb tethering (not Ethernet) is not possible because of limitations of devices... I think however usb Ethernet should work also on quest

**Comment by Woohaaa** [comment]
Controller vibration losing via ALVR？I think polygraphene is working hard to solve many bugs.

**Comment by FricoRico** [comment]
> I'm currently digging through my Quest's filesystem to try and find the models for the new controllers. With some work they could be adapted to SteamVR's rendermodel system.

Look no further, I have the models right here for you:
[OculusTouchForQuestAndRiftSControllers.zip](https://github.com/polygraphene/ALVR/files/3219252/OculusTouchForQuestAndRiftSControllers.zip)

**Comment by lvoogdt** [comment]
I did some digging around in the code, and found that it's possible to render the cv1 models for left and right controllers. "oculus_cv1_controller_left" and "oculus_cv1_controller_right"

@polygraphene I also saw that device.HasTouchController returns false for the quest (in ServerConfig). It's because DeviceSubType is 0, while it should be 3 (HelloListener.ALVR_DEVICE_SUBTYPE_OCULUS_MOBILE_QUEST)

**Comment by polygraphene** [comment]
> I did some digging around in the code, and found that it's possible to render the cv1 models for left and right controllers. "oculus_cv1_controller_left" and "oculus_cv1_controller_right"
> 
> @polygraphene I also saw that device.HasTouchController returns false for the quest (in ServerConfig). It's because DeviceSubType is 0, while it should be 3 (HelloListener.ALVR_DEVICE_SUBTYPE_OCULUS_MOBILE_QUEST)

I forgot to add detection for Quest on client side. After I added this commit (https://github.com/polygraphene/ALVRClient/commit/c3885d35a1d085cd6d50ce68fe896a8d2f82c088),
proper render model appeared with your PR.
Thank you!

**Comment by Irastris** [comment]
Any chance we can get the rendermodel redirected to a setting in the GUI so that it could be modified without having to recompile the server?

**Comment by polygraphene** [comment]
@Irastris 
I already fixed render model issue thanks to @lvoogdt 's PR #305.
alpha3 will include this patch.

**Comment by moallan03** [comment]
Will alpha 3 include a fix for controller buttons?

**Comment by Irastris** [comment]
The desire to allow specifying rendermodel in the GUI is mostly to make testing my Quest controller rendermodel (thanks @FricoRico!) testing a bit easier. I can just backup the oculus_cv1_controller_left/right folders and place it there though.

**Comment by polygraphene** [comment]
Button issue was partially fixed. Still has strange behavior.
https://github.com/polygraphene/ALVR/releases/tag/v2.4.0-alpha3

**Comment by Irastris** [comment]
First attempt at bringing the Quest Touch controller into SteamVR: https://i.imgur.com/JyGcudD.jpg

Would it perhaps be easier to wait for SteamVR to officially add rendermodels for the Rift S and then use those?

**Comment by VRtinker** [comment]
> Button issue was partially fixed. Still has strange behavior.
> https://github.com/polygraphene/ALVR/releases/tag/v2.4.0-alpha3

Hi @polygraphene,
As usual, outstanding efforts on display! I just wanted to report that despite the fact that now the models shown are no longer Vive wands, their orientation seems to be have kept the same. You'll see immediately what I'm saying when you put them next to each other. They overlap, which make the issue obvious, but the pointer as well don't shoot the beam in the direction you would expect. I quickly tried to fire up Eleven Table Tennis, and the Vive wands seem to make a comeback outside of SteamVR. The left trigger was not responsive, so I could spawn a ball to test the game itself.

Right now, the controller model-swap seems more cosmetic than anything else, I guess that's what I'm trying to say.

I have no doubt you'll sort this one out as well!

さよなら

**Comment by AnalogMan151** [comment]
This works pretty much perfectly! Controllers look like Oculus controllers, the Joystick simulates the touchpad for Vive, grip buttons work, face buttons work, the only button I couldn't get working in-game was the triggers where they worked before. When I pull them they react on the model and it works in SteamVR home but in-game they just don't seem to register. Thinking it may be a mapping issue.

**Comment by Irastris** [comment]
I've figured out how to make a reactive rendermodel, and am in progress on the left Touch controller. @VRtinker I also figured out your mapping issue. With my rendermodel, it's near if not 1:1. Once it's uploaded, you can confirm by peeking through the HMD's nose gap.

**Comment by DannyDan** [comment]
Really impressive guys. It runs very smooth. I think only the Buttons need some more work, and then its alrerady usable. 👍

**Comment by moallan03** [comment]
I installed alvr correctly but my headset just wont connect, any ideas?

**Comment by lvoogdt** [comment]
@moallan03 disable your firewall

**Comment by lvoogdt** [comment]
> I've figured out how to make a reactive rendermodel, and am in progress on the left Touch controller.

Great work, ave you found any good resources on this?

**Comment by Irastris** [comment]
> Great work, have you found any good resources on this?

Unfortunately no, I resorted to repeatedly reading over the .json file for the CV1 rendermodel and coming to an understanding from that.

Here are the left and right Quest Touch rendermodels. Backup and replace the folders located at `[steam install dir]\steamapps\common\SteamVR\resources\rendermodels` with these to enjoy Quest controllers in SteamVR! Tap your controllers together and see how accurate the mapping is ;)

https://anonfile.com/U9u2Q1sbnc/QuestTouchRendermodels_zip

The only thing missing is a reactive joystick. Since ALVR doesn't currently show that properly with the CV1 rendermodel, I had no basis for this Quest rendermodel and opted to not bother for now.

**Comment by VRtinker** [comment]
> > Great work, have you found any good resources on this?
> 
> Unfortunately no, I resorted to repeatedly reading over the .json file for the CV1 rendermodel and coming to an understanding from that.
> 
> Here are my left and right Quest Touch rendermodels. Backup and replace the folders located at `[steam install dir]\steamapps\common\SteamVR\resources\rendermodels` with these to enjoy Quest controllers in SteamVR! Tap your controllers together and see how accurate the tracking is ;)
> 
> https://files.catbox.moe/6uhfa5.zip
> 
> The only thing missing is a reactive joystick. Since ALVR doesn't currently show that properly with the CV1 rendermodel, I had no basis for this Quest rendermodel and opted to not bother for now.
> 
> @polygraphene You're welcome to use these in your driver (credit me and @FricoRico perhaps?) so that end users don't need to do this hacky file replacement.

Hey, I haven't tried your files yet, but from the look of it, it seems they would replace the original touch controller models. I own a CV1 and it seems that this "fix" would cause the original touch controllers to be misrepresented. Wouldn't that be the case?

**Comment by Irastris** [comment]
> Hey, I haven't tried your files yet, but from the look of it, it seems they would replace the original touch controller models. I own a CV1 and it seems that this "fix" would cause the original touch controllers to be misrepresented. Wouldn't that be the case?

Yes, hence my closing statement to @polygraphene. Once they were implemented in the ALVR driver instead, the replacement of the CV1 models would no longer be necessary. For now, this is a fix mostly intended for developers/early testers.

**Comment by moallan03** [comment]
Which one?

**Comment by VRtinker** [comment]
@Irastris Sorry, I misread than. It sounded like you were just offering him to use the models you guys created. Instead he needs to target an entirely new set which currently doesn't exist. I'd be content if the orientation were to be correct, despite the old touch controller being displayed though. Aside from the reverse ring, they are literally the same in terms of orientation of the handle/buttons, etc.

**Comment by Irastris** [comment]
> Sorry, I misread than. It sounded like you were just offering him to use the models you guys created. Instead he needs to target an entirely new set which currently doesn't exist.

They _do_ exist, and that _is_ what I'm offering. In the meantime however, early adopters can test it out using my instructions.

**Comment by VRtinker** [comment]
They don't exist in SteamVR, that's what I meant. I have Vive wands and WMR controller models in my rendermodels folder even though I don't own any of those. At any rate, I apologize for my audacity

**Comment by moallan03** [comment]
Just got it to work thanks

**Comment by Irastris** [comment]
Unfortunately it seems there's more work involved than just slapping my models into `ALVR/driver/resources/rendermodels` which is again beyond my knowledge level. We will need to wait for @polygraphene or @Ivoogdt to properly implement them.

**Comment by jarodhanko** [comment]
I've been attempting to use quest with echovr (just to see what it can do).  Alpha 3 fixes the button mappings for me, however the controller angle is off by 45 degrees or so.  Also throwing the disc is impossible for some reason.  I dont know if it is missing imu data that it relies on, or if there is something else going on.

**Comment by rathugo** [comment]
Tested Alpha 3 in Onward VR and the controls did not work, I could see the models of the touch controllers showing the buttons working but Onward was not detecting them. I could only grip and use the A button to select things. Played around in the SteamVR Home and it seemed to do better, could teleport and snapturn and it seemed to be detecting the controls. Will need to figure out why some games are not detecting the controls but is promising to see good performance tracking and latency wise this early on.

**Comment by leo60228** [comment]
@AnalogMan151 shouldn't it just be whatever the rift uses? the button layout is the same

**Comment by AnalogMan151** [comment]
@leo60228 ALVR doesn't identify the Oculus Touch controllers as Touch controllers to SteamVR, it tells SteamVR they are Vive controllers. So every game you load up is configured to expect Vive inputs. ALVR button maps the real Touch controls to these fake Vive controllers but these remaps aren't picked up properly by all games it seems.

**Comment by Irastris** [comment]
I don't think that's right, see [#L183](https://github.com/Irastris/ALVR/blob/2d715f00255b2612ba13458f3f21788b1f217208/ALVR/ServerConfig.cs#L183) of ServerConfig.cs. It's specifying that the Touch controllers are infact Touch controllers.

I did notice that going into the SteamVR Input Configuration for each game revealed that it was still assigning them the Vive profile by default though. You might try switching to the Touch profile and seeing what happens then.

**Comment by AnalogMan151** [comment]
That's just for loading the Oculus Touch models in SteamVR, the games still behave like you're using Vive controllers. As for setting the Controller type in the profile settings, A) That's only an option for games in your Steam Library and B) doesn't work anyway.

**Comment by juniortivotyro** [comment]
If it is possible, it should tell steamvr that the controllers are Oculus touch controllers. Because it tells games they are Vive controllers, it makes games such as H3VR unplayable, even if button mappings were correct, due to the angle of the controllers.

**Comment by AnalogMan151** [comment]
Games not properly detecting the Touch controllers may be due to how the game checks for which controller is connected. The suggested method is by asking SteamVR, but most older games go by Oculus's Documentation which states the following:
```
When properly configured and connected with your operating system, the Oculus Touch Controllers appear in the list returned by UnityEngine.Input.GetJoystickNames() as Oculus Touch - Left and Oculus Touch - Right, and the Oculus Remote appears as Oculus Remote. 
```
Or via this method:
```
IVRSystem.GetStringTrackedDeviceProperty(Prop_ManufacturerName_String)
```
which returns either HTC or Oculus. Since ServerConfig.cs sets the Manufacturer and model of the controllers to ALVR, it's possible the games are not able to properly detect the controller type and are defaulting to Vive. Setting the Manufacturer to Oculus and the Joystick name to Oculus Touch - Left / Right then games may correctly detect the controllers and not need any button mapping.

**Comment by Irastris** [comment]
If only I knew how to setup a development environment, I'd try that change.

**Comment by juniortivotyro** [comment]
With that being the case, and with no prior knowledge, it seems like vive controllers are a fallback, which makes sense. I wonder if button mappings would also magically be fixed if this was implemented, assuming the quest controllers use the same button layout as the rift touch controllers in software.

**Comment by lvoogdt** [comment]
@Irastris  I made i pull request with your controllers in it. It only show them when the app hasnt replaced it with its own controllers. Steam home for example has it's own controller where the angle is still off. We can circumvent this by adjusting controllerType to something other than 'oculus_touch' but then the buttons arent working anymore.

**Comment by FricoRico** [comment]
> If only I knew how to setup a development environment, I'd try that change.

Setting up is not terribly hard (I did it) but it requires you to install CUDA SDK 9.3.2 and FreePIE as it taps in on these SDKs through hardcoded paths. That is my biggest gripe with the current setup, wish FreePIE and CUDA library would be available through NUGET but sadly they aren't.

**Comment by lvoogdt** [comment]
@polygraphene All buttons seem to work, the trigger buttons seems to work, but it's still not registered by some apps (google earth and xplane for example).

**Comment by DannyDan** [comment]
Yepp, I tested some games yesterday, and in any game at least one needed button was not mapped.

**Comment by polygraphene** [comment]
@Irastris
Thank you for great work!
Can I put your model data into ALVR repository?
In addition, I want to rename your model to oculus_quest_controller_left(right) in order to avoid misunderstanding.
Is it okay?

> Unfortunately it seems there's more work involved than just slapping my models into `ALVR/driver/resources/rendermodels` which is again beyond my knowledge level. We will need to wait for @polygraphene or @Ivoogdt to properly implement them.

It seems that RenderModelName should be full path of model folder.
I will fix it on the next alpha release.

**Comment by polygraphene** [comment]
Trigger problem will also be fixed. https://github.com/polygraphene/ALVRClient/commit/871ba4241e5759b14733d134cf8b6faf94fbdb08

**Comment by Irastris** [comment]
> Can I put your model data into the ALVR repository?

@polygraphene Go right ahead! I already modified them to the oculus_quest_controller_[left|right] naming scheme for you, as well as did some cleanup (removed duplicate MTL files). Let me know if there's any issues.

https://files.catbox.moe/1zbqvu.zip

As soon as the joystick is functioning on the CV1 model, I'll make that reactive on my Quest model and make a pull request for those changes. Can we get a "force CV1 model" toggle in the GUI so that I'll be able to see when that's taken care of?

**Comment by SweViver** [comment]
Hi @polgraphene I would like to thank you so much for making this software. You are doing an amazing job. I spent this weekend testing ALVR and Im seriously shocked how great it works. Really impressive!

Keep up your great job man, this could get huge!
By the way, I made a video on my channel about ALVR testing it with Projeft Cars2, Onward, DCS, X Plane 11 and others. I hope it will help bringing some more attention to your great project!
https://youtu.be/gtmJInS7RxU

Cheers and thanks again!
/Martin

**Comment by YeetSkillet** [comment]
Idk if its my connection or not but my quest has really high latency rates and constant stuttering whenever i use alvr (around 60ms) if i lower the bitrate the latency is better but my resolution drops. I have a 2.4Ghz router

**Comment by rathugo** [comment]
> I have a 2.4Ghz router
@YeetSkillet ALVR will benefit a lot from a 5Ghz connection.

**Comment by machenmusik** [comment]
re: https://github.com/polygraphene/ALVR/issues/287#issuecomment-496002766

see #313 - for me, that commit did not resolve the issue

**Comment by polygraphene** [comment]
New alpha.
https://github.com/polygraphene/ALVR/releases/tag/v2.4.0-alpha4

**Comment by polygraphene** [comment]
@YeetSkillet 
60ms latency is not bad. I have same latency.
What's the stuttering problem? Do you experience stutter animation? Or head tracking?

**Comment by d4nielrx** [comment]
Will usb c tethering ever work? Surely itd be faster? Ivry says its possible dunno how true that is

**Comment by Irastris** [comment]
iVRy also hasn't released even a demo video at the least, so I'd take that with a grain of salt for now.

**Comment by aebrett** [comment]
@polygraphene, this is the build I’ve been waiting for! I’ve tried the previous alphas, and this is the first that (so far) works perfectly with Google Earth VR. Before my quest arrived, I was obsessively refreshing the Amazon delivery status, but over the last few days I’ve been refreshing the ALVR commits page waiting for new builds. Thanks so much for your work, and the frequent release builds since the quest was released; I’ve sent a small paypal donation in thanks.

**Comment by peterthrees** [comment]
@polygraphene I tried to start the app in windows, but i'm getting an error:
https://www.screencast.com/t/hiFCluux3I  (i used the EXE to install)

"The system cannot find the file specified"

This is are the details:

See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at ALVR.Utils.ExecuteProcess(String path, String args)
   at ALVR.DriverInstaller.GetDriverList()
   at ALVR.DriverInstaller.RemoveOtherDriverInstallations()
   at ALVR.Launcher.Launcher_Load(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at MetroFramework.Forms.MetroForm.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at MetroFramework.Forms.MetroForm.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)


************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3130.0 built by: NET472REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
----------------------------------------
ALVR
    Assembly Version: 1.1.0.0
    Win32 Version: 1.0.0.0
    CodeBase: file:///C:/Program%20Files/ALVR/ALVR.exe
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
MetroFramework
    Assembly Version: 1.2.0.3
    Win32 Version: 1.2.0.3
    CodeBase: file:///C:/Program%20Files/ALVR/MetroFramework.DLL
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3081.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
MetroFramework.Fonts
    Assembly Version: 1.2.0.3
    Win32 Version: 1.2.0.3
    CodeBase: file:///C:/Program%20Files/ALVR/MetroFramework.Fonts.DLL
----------------------------------------
Accessibility
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Accessibility/v4.0_4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll
----------------------------------------
Microsoft.CSharp
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Microsoft.CSharp/v4.0_4.0.0.0__b03f5f7f11d50a3a/Microsoft.CSharp.dll
----------------------------------------
System.Runtime.Serialization
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3130.0 built by: NET472REL1LAST_B
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Serialization/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Serialization.dll
----------------------------------------
System.Xml.Linq
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0 built by: NET472REL1
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml.Linq/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.Linq.dll
----------------------------------------
SMDiagnostics
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3130.0 built by: NET472REL1LAST_B
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/SMDiagnostics/v4.0_4.0.0.0__b77a5c561934e089/SMDiagnostics.dll
----------------------------------------
System.ServiceModel.Internals
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3130.0 built by: NET472REL1LAST_B
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.ServiceModel.Internals/v4.0_4.0.0.0__31bf3856ad364e35/System.ServiceModel.Internals.dll
----------------------------------------
System.Dynamic
    Assembly Version: 4.0.0.0
    Win32 Version: 4.7.3062.0
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Dynamic/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Dynamic.dll
----------------------------------------
Anonymously Hosted DynamicMethods Assembly
    Assembly Version: 0.0.0.0
    Win32 Version: 4.7.3130.0 built by: NET472REL1LAST_B
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/mscorlib/v4.0_4.0.0.0__b77a5c561934e089/mscorlib.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.

**Comment by polygraphene** [comment]
@peterthrees
Please check the installation of SteamVR.
You need to install SteamVR on C:\ drive.

**Comment by peterthrees** [comment]
@polygraphene, thanks I've fixed that, next I've allowed port 9944 in my  firewall. I get connection with my quest, but when i want to start the server, i'm receiving the following error:

Failed to initialize CEncoder. All VideoEncoder are not available. VCE: AMF Error 1. g_AMFFactory.Init(), NVENC: NvEnc NvEncoderD3D11 failed. Code=1 NvEncoder::LoadNvEncApi : NVENC library file is not found. Please ensure NV driver is installed at c:\src\alvr\alvr_server\nvencoder.cpp:70

Any idea how to solve this?

**Comment by polygraphene** [comment]
@peterthrees 
Please upgrade graphic driver.

**Comment by peterthrees** [comment]
@polygraphene Thanks, i was trying to do this on my work notebook. Maybe not a good idea for steam :) 
I've tried it on my more powerfull pc. I bought it to run oculus dk2 games. I have  AMD Radeon R9 200 gfx card, when I run the vr game from steam without my quest, it starts. When I use ALVR to run it on my quest, I get the same error issue again. The drivers are the latest ones. Used riftcat on my notebook before and thjat one worked but not that good.

Is ALVR maybe using something like what the app moonlight does? I know it only support gforce gfx cards. Thats why i used parsec instead.

**Comment by Irastris** [comment]
@peterthrees ALVR is currently 100% reliant on NVENC for encoding, which is unavailable on AMD cards. Apologies.

**Comment by peterthrees** [comment]
Ok thanks, ill try to find a second hand gforce card that support nvenc

**Comment by polygraphene** [comment]
Though ALVR supports Radeon GPU, R9 200 series may be a bit old for ALVR.
I tested only on RX 400 and 500 series.

**Comment by YeetSkillet** [comment]
@polygraphene

Stuttering in the headset tracking, the controllers for the most part work fine but the headset stuttering gives me nausea

**Comment by jarodhanko** [comment]
> Stuttering in the headset tracking, the controllers for the most part work fine but the headset stuttering gives me nausea

@YeetSkillet 

You will have to give more info than that.  Likely that is a problem with your network setup or the settings chosen in ALVR are too high for your network to handle.

Latency will depend on network setup (2.4 ghz, 5ghz, network speed, distance from access point) and application specific settings (resolution, bitrate).  You may need to adjust your settings in order to get low enough latency.  If you believe it is an issue with ALVR itself, you need to provide details on how you have things setup or how it can be reproduced.

1. Distance from route/access point
2. 2.4 ghz vs 5 ghz
3. how many other wifi networks are you competing with (how congested is the wifi channel that you are using)
4. Is there other traffic that may be using up bandwidth
5. What are your computer specs
6. What application(s) are you attempting to use

This is just an example, but no one will be easily able to help you without supporting information.

**Comment by peterthrees** [comment]
Who wants to know, there is a topic on resetera on ALVR and others:

https://www.resetera.com/threads/oculus-quest-steamvr-thread-what-can-5ghz-buy-you.119494/

**Comment by polygraphene** [comment]
> @polygraphene
> 
> Stuttering in the headset tracking, the controllers for the most part work fine but the headset stuttering gives me nausea

Please create new issue. And tell me your OS and GPU.

**Comment by crazysword1** [comment]
Thank you for your hard work. Just want to confirm one thing.

With all other factors being equal, does a higher bitrate translate to higher latency? What about if the router is sufficient to handle the higher bitrate? Does it take longer to encode and decode a higher bitrate stream?

Also for anyone using H.265, are you using the Pascal (Nvidia 1000 series) or Turing (2000 series) cards?

**Comment by FingrMastr** [comment]
Here is my config:
Intel i5 8400
16Gb ram 2666mhz
Nvidia GeForce 2070 8Gb hheavily overclocked
PC on 100Mbit Ethernet cabled connection to router
Oculus Go on cabled rj45 connection to router using an adapter Ethernet/microusb
ALVR stable (not alpha nor beta)
Maxed resolution on ALVR
50mbit bitrate
300kb buffer
Max latency: 70ms
Totally playable on Project Cars 2, stable 72 fps, no artifacts, on ultra settings.
SIMPLY FANTASTIC

**Comment by Cumulus7** [comment]
> Though ALVR supports Radeon GPU, R9 200 series may be a bit old for ALVR.
> I tested only on RX 400 and 500 series.
Just for your information:
I got GoogleEarth VR running yesterday using an AMD HD 7970  - (first generation GCN, same as the R9 280X). Works just fine.
Only problem here: resolution has to be set to 75% or lower :-(
Don't know why that may be.....

THX for your work - i think i might have to upgrade the gfx card after the summer.

**Comment by Thijzer** [comment]
Just had a successful install. Was up and running in minutes. I started with sideQuest. I would every beginner to do this route. Follow the setup on sideQuest instructions and don't rush. It helps you  enabling developer mode, install and drivers, install ALVR on the headset. Remaining things are the ALVR server setup and steam vr setup. 

So in order to make those work install steam vr in steam make sure to play as well.
Make a connection with the headset going to ALVR app in the headset and connect the headset on the server.
Last when everything is installed restart all mentioned apps or the computer. 
Start ALVR server up again make the connection and start the server. 
You should be good to go.

I have RX580 and streaming works just like steam link works.

Steam vr was never on my computer but it found my devices.

My WiFi connection is not optimal and it works well enough to make it work during setup. For playing I will make a guard zls close as possible to the WiFi router.

I also have a rental router from my provider so nothing custom or expensive.

**Comment by gtrak** [comment]
Got up and running with sidequest, oculus quest, GTX 1070.  I ran Dirt Rally with Revive.  The hand controllers were a complete mess (maybe button mapping?), so I disabled them and used a 3rd party bluetooth gamepad.  Streaming was smooth, tracking was stuttery.  Going to try setting TrackingFrameOffset to -1 if you 'experience jerky head-tracking' once I stop feeling motion sickness, probably the fault of my choice of game :-).

**Comment by gtrak** [comment]
TrackingFrameOffset didn't help. Forcing 60hz didn't help.  I noticed occasionally the tracking would be smooth for a second or two, but it was very rare.  I also noticed in the graphics settings of dirt rally, the refresh rate says 59, though I ran a benchmark and it didn't go higher than 72.40 FPS, which is suspicious.
